redbox-review/redbox.go

```diff
@@ -0,0 +1,293 @@
+package redbox
+
+import (
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/cgclever/redbox/s3box"
+)
+
+const defaultNManifests = 4
+
+var (
+	errShippingInProgress = fmt.Errorf("Cannot perform any action when shipping is in progress.")
+	errIncompleteArgs     = fmt.Errorf("Creating a redshift box requires a schema, table and an s3 bucket.")
+	errInvalidJSONInput   = fmt.Errorf("Only JSON inputs are supported.")
+	errBoxShipped         = fmt.Errorf("Cannot perform any actions, the box has been shipped.")
+	errNothingToShip      = fmt.Errorf("Cannot perform send, no data was packed.")
+)
+
+// Redbox manages piping data into Redshift.
+// An S3Box is used to manage data transport to S3 via Pack, after
+// which Ship commits the data to Redshift.
+type Redbox struct {
+	// Inheret mutex locking/unlocking
+	mt sync.Mutex
+
+	// schema is the schema of the destination
+	schema string
+
+	// table is the table name of the destination
+	table string
+
+	// awsKey is the AWS ACCESS KEY ID to the s3bucket
+	awsKey string
+
+	// awsPassword is the AWS SECRET ACCESS KEY to the s3Bucket
+	awsPassword string
+
+	// s3Bucket is the bucket storing our data
+	s3Bucket string
+
+	// s3Region is the location of the destination s3Bucket
+	s3Region string
+
+	// truncate indicates if we should truncate the destination table
+	truncate bool
+
+	// nManifests is the number of manifests to split streamed data into.
+	nManifests int
+
+	// s3Box manages the transport of data to Redshift
+	s3Box s3box.S3BoxAPI
+
+	// redshift is the direct redshift connection
+	redshift *sql.DB
+
+	// shippingInProgress indicates if a send is in progress
+	shippingInProgress bool
+
+	// shipped indicates if the box has been shipped
+	shipped bool
+}
+
+// NewRedboxOptions specifies the configuration for a new Redbox
+type NewRedboxOptions struct {
```

Comment:
Doesn't seem like a bad idea, given that it's terser. However, is this in line with other clients? I've seen examples, including `aws-sdk-go` and `Wag` where for a client method `Method` the argument is of the form `MethodInput` or `MethodOptions`.